drivers/gpio/gpio_max22190.c

```diff
@@ -0,0 +1,646 @@
+ * @param addr - Register value to which data is written.
+ * @param val - Value which is to be written to requested register.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int max22190_reg_transceive(const struct device *dev, uint8_t addr, uint8_t val, uint8_t rw)
+{
+	uint8_t crc;
+	int ret;
+
+	uint8_t local_rx_buff[MAX22190_MAX_PKT_SIZE] = {0};
+	uint8_t local_tx_buff[MAX22190_MAX_PKT_SIZE] = {0};
+
+	const struct max22190_config *config = dev->config;
+
+	struct spi_buf tx_buf = {
+				.buf = &local_tx_buff,
+				.len = config->pkt_size,
+				};
+	const struct spi_buf_set tx = {.buffers = &tx_buf, .count = 1};
+
+	struct spi_buf rx_buf = {
+				.buf = &local_rx_buff,
+				.len = config->pkt_size,
+				};
+	const struct spi_buf_set rx = {.buffers = &rx_buf, .count = 1};
+
+	if (config->crc_en & 0) {
+		rx_buf.len++;
+	}
+
+	local_tx_buff[0] = FIELD_PREP(MAX22190_ADDR_MASK, addr) |
+				FIELD_PREP(MAX22190_RW_MASK, rw & 0x1);
+	local_tx_buff[1] = val;
+
+	/* If CRC enabled calculate it */
+	if (config->crc_en) {
+		local_tx_buff[2] = max22190_crc(&local_tx_buff[0]);
+	}
+
+	/* write cmd & read resp at once */
+	ret = spi_transceive_dt(&config->spi, &tx, &rx);
+
+	if (ret) {
+		LOG_ERR("Err spi_transcieve_dt  [%d]\n", ret);
+		return ret;
+	}
+
+	/* if CRC enabled check readed */
+	if (config->crc_en & 0) {
+		crc = max22190_crc(&local_rx_buff[0]);
+		if (crc != (local_rx_buff[2] & 0x1F)) {
+			LOG_ERR("READ CRC ERR (%d)-(%d)\n", crc, (local_rx_buff[2] & 0x1F));
+			return -EINVAL;
+		}
+	}
+
+	/* always (R/W) get DI reg in first byte */
+	max22190_update_in_stat(dev, local_rx_buff[0]);
+
+	/* in case of writing register we get as second byte WB reg */
+	if (rw == MAX22190_WRITE) {
+		max22190_update_wb_stat(dev, local_rx_buff[1]);
+	} else {
+		/* in case of READ second byte is value we are looking for */
+		ret = local_rx_buff[1];
+	}
+
+	return ret;
+}
+
+/*
+ * @brief Register update function for MAX22190
+ *
+ * @param dev - MAX22190 device.
+ * @param addr - Register valueto wich data is updated.
+ * @param mask - Corresponding mask to the data that will be updated.
+ * @param val - Updated value to be written in the register at update.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int max22190_reg_update(const struct device *dev, uint8_t addr, uint8_t mask, uint8_t val)
+{
+	int ret;
+	uint32_t reg_val = 0;
+
+	ret = max22190_reg_transceive(dev, addr, 0, MAX22190_READ);
+
+	reg_val = ret;
+	reg_val &= ~mask;
+	reg_val |= mask & val;
+
+	return max22190_reg_transceive(dev, addr, reg_val, MAX22190_WRITE);
+}
+
+/*
+ * @brief Check FAULT1 and FAULT2
+ *
+ * @param dev - MAX22190 device
+ */
+static void max22190_fault_check(const struct device *dev)
+{
+	/* FAULT1 */
+	struct max22190_data *data = dev->data;
+
+	uint8_t _fault1 = max22190_reg_transceive(dev, MAX22190_FAULT1_REG, 0,
+							MAX22190_READ);
+
+	if (_fault1) {
+		memcpy(&data->fault1, &_fault1, 1);
```

Comment:
```suggestion
		data->fault1 = fault1;
```

Reply:
`data->fault1` is structure
To avoid memcpy I can wrap it in union ?